R/tm_g_ae_oview.R

```diff
@@ -285,34 +281,32 @@ srv_g_ae_oview <- function(id,
 
     output_q <- reactive({
       ANL <- data[[dataname]]() # nolint
-      validate(need(iv$is_valid(), "Misspecification error: please observe red flags in the encodings."))
-      validate(need(
-        is.factor(ANL[[input$arm_var]]),
-        "Selected arm variable needs to be a factor."
-      ))
-      validate(need(input$flag_var_anl, "Please select at least one flag."))
-      iv_comp <- shinyvalidate::InputValidator$new()
-      iv_comp$add_rule("arm_trt", shinyvalidate::sv_not_equal(
+
+      teal::validate_has_data(ANL, min_nrow = 10, msg = sprintf("%s has not enough data", dataname))
+
+      # set up and enable input validator(s)
+      iv <- shinyvalidate::InputValidator$new()
+      iv$add_rule("arm_var", shinyvalidate::sv_required(
+        message = "Arm Variable is required"))
+      iv$add_rule("arm_var", ~ if (!is.factor(ANL[[req(.)]]))
+        "Arm Var must be a factor variable")
+      iv$add_rule("arm_var", ~ if (length(unique(ANL[[req(.)]])) < 2)
+        "Selected Arm Var has not enough treatments to compare")
+      iv$add_rule("flag_var_anl", shinyvalidate::sv_required(
+        message = "At least one Flag is required"))
+      iv$add_rule("arm_trt", shinyvalidate::sv_not_equal(
         input$arm_ref,
-        message_fmt = "Must not be equal to Control"
-      ))
-      iv_comp$add_rule("arm_ref", shinyvalidate::sv_not_equal(
+        message_fmt = "Control and Treatment must be different"))
+      iv$add_rule("arm_ref", shinyvalidate::sv_not_equal(
         input$arm_trt,
-        message_fmt = "Must not be equal to Treatment"
-      ))
-      iv_comp$enable()
+        message_fmt = "Control and Treatment must be different"))
+      iv$enable()
 
-      validate(need(iv_comp$is_valid(), "Misspecification error: please observe red flags in the encodings."))
-      validate(need(nlevels(ANL[[input$arm_var]]) > 1, "Arm needs to have at least 2 levels"))
-      validate_has_data(ANL, min_nrow = 10)
-      if (all(c(input$arm_trt, input$arm_ref) %in% ANL[[input$arm_var]])) {
-        iv_an <- shinyvalidate::InputValidator$new()
-        iv_an$add_rule("arm_ref", shinyvalidate::sv_in_set(set = ANL[[input$arm_var]]))
-        iv_an$add_rule("arm_trt", shinyvalidate::sv_in_set(set = ANL[[input$arm_var]]))
-        iv_an$enable()
-        validate(need(iv_an$is_valid(), "Misspecification error: please observe red flags in the encodings."))
-      }
-      validate(need(all(c(input$arm_trt, input$arm_ref) %in% unique(ANL[[input$arm_var]])), "Plot loading"))
+      # collate validator messages
+      gather_fails(iv)
+
+      validate(need(all(c(input$arm_trt, input$arm_ref) %in% unique(ANL[[input$arm_var]])),
+                    "Plot loading"))
```

Comment:
Can we change this (I know you didn't add it in this PR but it's clearly wrong...)

Reply:
Can you explain how it is wrong? I assumed it's a DYI spinner as it does that job rather well.

Reply:
The plot isn't loading - there's a validation error!

Reply:
I thought that was the point. 

Fair enough, fixed. 

By the way, this is one of several redundant validations. One input selects a variable and that causes another input to present this variable's available values as choices. The value selected in the second input cannot but be a element of the variable selected in the first one. Why validate it at all?

Reply:
This can be turned into a meaningful message after all. If you filter too much, the condition WILL be broken.

I will have a look at all modules.